### PR TITLE
[API][Checkout]checkout without logged in account

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         |  master
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: features/user/managing_customers/editing_customer.feature [hunk 1]
```diff
@@ -46,3 +46,12 @@ Feature: Editing a customer
         And I save my changes
         Then I should be notified that it has been successfully edited
         And this customer should have "Retail" as their group
+
+    @ui
+    Scenario: Verifying customer account
```
Comment (GSadee) [inline features/user/managing_customers/editing_customer.feature:51]: Why is this scenario in this PR? It looks like a completely separate case

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 2]
```diff
@@ -224,6 +225,14 @@ public function iShouldBeOnTheCheckoutShippingStep(): void
         Assert::same($value, OrderCheckoutStates::STATE_ADDRESSED);
     }
 
+    /**
+     * @Given I complete the shipping step
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:229]:
```suggestion
     * @When I complete the shipping step
```

File: src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml [hunk 3]
```diff
@@ -12,11 +12,11 @@ default:
                 - sylius.behat.context.transform.cart
                 - sylius.behat.context.transform.lexical
                 - sylius.behat.context.transform.order
-                - sylius.behat.context.transform.payment
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml:15]: Why did you move this service?

File: src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml [hunk 4]
```diff
@@ -26,10 +26,13 @@ default:
                 - sylius.behat.context.setup.product
                 - sylius.behat.context.setup.shipping
                 - sylius.behat.context.setup.shop_api_security
+                - sylius.behat.context.setup.customer
+                - sylius.behat.context.setup.payment
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml:30]: Services should be sorted in the alphabetical order

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 5]
```diff
@@ -224,6 +225,14 @@ public function iShouldBeOnTheCheckoutShippingStep(): void
...
+     */
+    public function iCompleteTheShippingStep(): void
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:231]: In line 54 we already have one empty step. it can be merged.

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php [hunk 6]
```diff
@@ -64,24 +70,24 @@ public function __invoke(AddressOrder $addressOrder): OrderInterface
         );
 
         /** @var CustomerInterface|null $customer */
-        $customer =  $order->getCustomer();
+        $customer = $this->customerRepository->findOneBy(['email' => $addressOrder->email]);
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php:73]: i am not sure about this logic. Here we need a customer which is related to order, no customer from database, because in DB the customer can exist but he can checkout his order without login. Your suggestion makes this case impossible :(

File: src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml [hunk 7]
```diff
@@ -12,11 +12,11 @@ default:
                 - sylius.behat.context.transform.cart
                 - sylius.behat.context.transform.lexical
                 - sylius.behat.context.transform.order
-                - sylius.behat.context.transform.payment
                 - sylius.behat.context.transform.product
                 - sylius.behat.context.transform.shared_storage
                 - sylius.behat.context.transform.shipping_method
                 - sylius.behat.context.transform.user
+                - sylius.behat.context.transform.payment
```
Comment (AdamKasp) [inline src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml:19]: wrong order

File: features/user/managing_customers/editing_customer.feature (hunk 1 above)
Comment (AdamKasp) [inline features/user/managing_customers/editing_customer.feature:51]: why did you add a new `@ui` scenario?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php (hunk 6 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php:73]: I guess that only if there is no customer assigned to an order, we should search customer in the repository, because in case that customer is logged in, the email could be passed as null.

File: src/Sylius/Behat/Context/Transform/CartTokenContext.php [hunk 8]
```diff
@@ -43,6 +43,7 @@ public function __construct(
 
     /**
      * @Transform /^(cart)$/
+     * @Transform :cart
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Transform/CartTokenContext.php:46]: where did you use this new transformer?

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/AddressOrderHandlerSpec.php [hunk 9]
```diff
@@ -133,6 +139,49 @@ function it_handles_addressing_an_order_for_logged_in_shop_user(
...
+
+        $customerRepository->findOneBy(['email' => 'r2d2@droid.com'])->willReturn($customer);
+
+        $order->getCustomer()->shouldBeCalled();
+        $order->setCustomer($customer)->shouldBeCalled();
+        $order->setBillingAddress($billingAddress)->shouldBeCalled();
+        $order->setShippingAddress($shippingAddress)->shouldBeCalled();
+
+        $stateMachineFactory->get($order, OrderCheckoutTransitions::GRAPH)->willReturn($stateMachine);
+        $stateMachine->can('address')->willReturn(true);
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/AddressOrderHandlerSpec.php:179]: can you use here this const value?
```suggestion
        $stateMachine->can(OrderCheckoutTransitions::TRANSITION_ADDRESS)->willReturn(true);
```

File: src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml [hunk 10]
```diff
@@ -28,6 +28,7 @@ default:
                 - sylius.behat.context.setup.shipping
                 - sylius.behat.context.setup.shop_api_security
                 - sylius.behat.context.setup.user
+                - sylius.behat.context.setup.customer
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml:31]: Try to keep services in alphabetical order

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php [hunk 11]
```diff
@@ -65,16 +71,9 @@ public function __invoke(AddressOrder $addressOrder): OrderInterface
...
-            /** @var CustomerInterface $customer */
-            $customer = $this->customerFactory->createNew();
-            $customer->setEmail($addressOrder->email);
-
-            $this->manager->persist($customer);
+        $customer = $this->provideCustomer($order, $addressOrder->email);
 
+        if (null === $order->getCustomer()) {
             $order->setCustomer($customer);
         }
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php:78]:
```suggestion
        if (null === $order->getCustomer()) {
            $order->setCustomer($this->provideCustomerByEmail($addressOrder->email));
        }
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php [hunk 12]
```diff
@@ -83,6 +82,27 @@ public function __invoke(AddressOrder $addressOrder): OrderInterface
...
+        if (null === $customer) {
+            Assert::notNull($email, sprintf('Visitor should provide an email.'));
+
+            /** @var CustomerInterface $customer */
+            $customer = $this->customerFactory->createNew();
+            $customer->setEmail($email);
+            $this->manager->persist($customer);
+        }
+
+        return $customer;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php:106]:
```suggestion
    private function provideCustomerByEmail(?string $email): CustomerInterface
    {
        Assert::notNull($email, sprintf('Visitor should provide an email.'));
        
        $customer = $this->customerRepository->findOneBy(['email' => $email]);
        if (null !== $customer) {
            return $customer;
        }

         /** @var CustomerInterface $customer */
         $customer = $this->customerFactory->createNew();
         $customer->setEmail($email);

        return $customer;
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php [hunk 13]
```diff
@@ -83,6 +82,27 @@ public function __invoke(AddressOrder $addressOrder): OrderInterface
 
         $stateMachine->apply(OrderCheckoutTransitions::TRANSITION_ADDRESS);
 
+        $this->manager->persist($order);
+        $this->manager->flush();
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php:86]: Did you check if it is necessary?

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 14]
```diff
@@ -58,7 +58,6 @@ public function __construct(
     /**
      * @Given I am at the checkout addressing step
      * @When I complete the shipping step
-     * @When I complete the payment step
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:61]: Why did you remove that?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php [hunk 15]
```diff
@@ -64,25 +70,33 @@ public function __invoke(AddressOrder $addressOrder): OrderInterface
...
         return $order;
     }
+
+    private function provideCustomerByEmail(?string $email): CustomerInterface
+    {
+        Assert::notNull($email, sprintf('Visitor should provide an email.'));
+
+        $customer = $this->customerRepository->findOneBy(['email' => $email]);
+        if (null === $customer) {
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Checkout/AddressOrderHandler.php:93]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/AddressOrderHandlerSpec.php [hunk 16]
```diff
@@ -23,18 +23,20 @@
...
 {
     function let(
         OrderRepositoryInterface $orderRepository,
+        CustomerRepositoryInterface $customerRepository,
         FactoryInterface $customerFactory,
         ObjectManager $manager,
         StateMachineFactoryInterface $stateMachineFactory
     ): void {
-        $this->beConstructedWith($orderRepository, $customerFactory, $manager, $stateMachineFactory);
+        $this->beConstructedWith($orderRepository,$customerRepository, $customerFactory, $manager, $stateMachineFactory);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Checkout/AddressOrderHandlerSpec.php:39]:
```suggestion
        $this->beConstructedWith($orderRepository, $customerRepository, $customerFactory, $manager, $stateMachineFactory);
```

**Comment by GSadee** [comment]
Thank you, @arti0090! :1st_place_medal:
